src/validation_functions.jl

```diff
@@ -0,0 +1,144 @@
+end
+
+function get_limits(valid_range::String, ps_struct::PowerSystemType)
+    #Gets min and max values from activepowerlimits for activepower, etc.
+    limits = getfield(ps_struct, Symbol(valid_range))
+    return limits
+end
+
+function get_limits(valid_range::Dict, unused::PowerSystemType)
+    #Gets min and max value defined for a field, e.g. "valid_range": {"min":-1.571, "max":1.571}.
+    limits = (min = valid_range["min"], max = valid_range["max"])
+    return limits
+end
+
+function validate_range(::String, valid_info::ValidationInfo, field_value)
+    #Validates activepower against activepowerlimits, etc.
+    is_valid = true
+    if !isnothing(valid_info.limits)
+        is_valid = check_limits_impl(valid_info, field_value)
+    end
+    return is_valid
+end
+
+function validate_range(::Dict, valid_info::ValidationInfo, field_value)
+    return check_limits(valid_info.field_type, valid_info, field_value)
+end
+
+function check_limits(::Type{T}, valid_info::ValidationInfo, field_value) where T <: Union{Nothing, Float64}
+    #Validates numbers.
+    return check_limits_impl(valid_info, field_value)
+end
+
+function check_limits(::Type{T}, valid_info::ValidationInfo, field_value) where T <: Union{Nothing, NamedTuple}
+    #Validates up/down, min/max, from/to named tuples.
+    @assert length(field_value) == 2
+    result1 = check_limits_impl(valid_info, field_value[1])
+    result2 = check_limits_impl(valid_info, field_value[2])
+    return result1 && result2
+end
+
+function check_limits_impl(valid_info::ValidationInfo, field_value)
+    is_valid = true
+    action_function = get_validation_action(valid_info.field_descriptor)
+    if (!isnothing(valid_info.limits.min) && field_value < valid_info.limits.min) ||
+        (!isnothing(valid_info.limits.max) && field_value > valid_info.limits.max)
+        is_valid = action_function(valid_info, field_value)
+    end
+    return is_valid
+end
+
+function get_validation_action(field_descriptor::Dict)
+    action = get(field_descriptor, "validation_action", "error")
+    if action == "warn"
+        action_function = validation_warning
+    elseif action == "error"
+        action_function = validation_error
+    else
+        error("Invalid validation action $action")
+    end
+    return action_function
+end
+
+function validation_warning(valid_info::ValidationInfo, field_value)
+    valid_range = valid_info.field_descriptor["valid_range"]
+    field_name = valid_info.field_descriptor["name"]
+    @warn "Invalid range" valid_info.struct_name field_name field_value valid_range valid_info.ps_struct
+    return true
+end
+
+function validation_error(valid_info::ValidationInfo, field_value)
+    valid_range = valid_info.field_descriptor["valid_range"]
+    field_name = valid_info.field_descriptor["name"]
+    @error "Invalid range" valid_info.struct_name field_name field_value valid_range valid_info.ps_struct
+    return false
+end
+
+#Could be called from PowerSimulations.
```

Comment:
nit:  Should probably have a formal docstring.  And it shouldn't reference PowerSimulations. I would vote for it being exported, but @claytonpbarrows and @jd-lara might disagree.

Reply:
I think it should be exported as well.